lib/puppet/module.rb

```diff
@@ -456,7 +456,7 @@ def subpath(type)
 
   def assert_validity
     if !Puppet::Module.is_module_directory_name?(@name) && !Puppet::Module.is_module_namespaced_name?(@name)
-      raise InvalidName, "Invalid module name #{@name}; module names must be alphanumeric (plus '-'), not '#{@name}'"
+      raise InvalidName, "Invalid module name #{@name}; module names should only contain lowercase letters, numbers, and underscores and must begin with a lowercase letter."
```

Comment:
From looking at `is_module_directory_name?` and `is_module_namespaced_name?`, it looks like `@name` must match one of either `/^[a-z][a-z0-9_]*$/` or `/^[a-zA-Z0-9]+[-][a-z][a-z0-9_]*$/` in order to be considered valid.
So we can have one of either
- a name that must start with a lower case letter and can only contain lower case letters, numbers, or underscores (a valid module name)
- a name that contains one or more lower case letters, upper case letters, or numbers followed by a dash, a lower case letter, and then either lower case letters, numbers or underscores (a valid namespaced module name)

I would update this to something like 
```
raise InvalidName, _(<<ERROR_STRING).chomp % { name: @name }
Invalid module name '%{name}'; module names must match either an installed
module name (ex. modulename) or a namespaced module name
(ex. author-modulename). That is, it must match either the expression
/^[a-z][a-z0-9_]*$/ or /^[a-zA-Z0-9]+[-][a-z][a-z0-9_]*$/
ERROR_STRING
```

We also need to add `_()` around the error message in order to translate the string at runtime. I've done that in the example above. You can read more about that at https://github.com/puppetlabs/puppet/blob/d068a9d9f8f6322ed5d403d99fb56f2b3f1b626f/CONTRIBUTING.md#writing-translatable-code